Tests/Marketplace/upload_packs.py

```diff
@@ -924,10 +929,163 @@ def get_images_data(packs_list: list):
+    if not task_status:
+        pack.status = PackStatus.FAILED_SIGNING_PACKS.name
+        pack.cleanup()
+        return False
+    task_status, _ = pack.zip_pack()
+    if not task_status:
+        pack.status = PackStatus.FAILED_ZIPPING_PACK_ARTIFACTS.name
+        pack.cleanup()
+        return False
+    return True
+
+
+def get_all_packs(packs_dict, extract_destination_path, id_set_path, marketplace):
+    """
+    Collect all packs from the id_set that are not in packs_dict
+    """
+    if not id_set_path or not os.path.isfile(id_set_path):
+        return packs_dict
+    packs_dict = dict(packs_dict)
+    with open(id_set_path) as f:
+        id_set_packs = json.load(f).get('Packs', [])
+    for pack_name in id_set_packs.keys():
+        if pack_name not in packs_dict:
+            pack = Pack(pack_name, os.path.join(extract_destination_path, pack_name), marketplace)
+            packs_dict[pack_name] = pack
+    return packs_dict
+
+
+def upload_packs_with_dependencies_zip(extract_destination_path, packs_dependencies_mapping, signature_key,
+                                       storage_bucket, storage_base_path, id_set_path, packs_list, marketplace):
+    """
+    Uploads packs with mandatory dependencies zip for all packs
+    """
+    logging.info("Starting to collect pack with dependencies zips")
+    packs_dict = {pack.name: pack for pack in packs_list}
+    packs_dict = get_all_packs(packs_dict, extract_destination_path, id_set_path, marketplace)
+    full_deps_graph: dict = {}
+    try:
+        for pack in packs_dict.values():
+            logging.info(f"Collecting dependencies of {pack.name}")
+            pack_with_dep_path = os.path.join(pack.path, "with_dependencies")
+            zip_with_deps_path = os.path.join(pack.path, pack.name + "_with_dependencies.zip")
+            upload_path = os.path.join(storage_base_path, pack.name, pack.name + "_with_dependencies.zip")
+            Path(pack_with_dep_path).mkdir(parents=True, exist_ok=True)
+            full_deps_graph = map_pack_dependencies_graph(pack.name, packs_dependencies_mapping, full_deps_graph)
+            if pack.name not in full_deps_graph:
+                logging.error(f"Skipping dependencies collection for {pack.name}. missing from full_deps_graph")
+                continue
+            pack_deps = full_deps_graph[pack.name]
+            if not (pack.zip_path and os.path.isfile(pack.zip_path)):
+                if not prepare_and_zip_pack(pack, signature_key):
+                    logging.warning(f"Skipping dependencies collection for {pack.name}. Failed zipping")
+                    continue
+            shutil.copy(pack.zip_path, os.path.join(pack_with_dep_path, pack.name + ".zip"))
+            for dep_name in pack_deps:
+                if dep_name not in packs_dict:
+                    dep_pack = Pack(dep_name, os.path.join(extract_destination_path, dep_name), marketplace)
+                    packs_dict[dep_name] = dep_pack
```

Comment:
In python that's not an issue since the iterable returned in from `pack_deps` should be a copy, but it is a bad practice.
I'll take care of this as part of the changes we discussed.